### PR TITLE
Implement ListWalletAddresses function

### PR DESCRIPTION
**Pull request by jalavosus**
## Changes

ListWalletAddresses (implemented in lnwallet/interface.go,
lnwallet/btcwallet/btcwallet, and lnrpc/walletrpc/walletkit) allows for
querying WalletKit to retrieve all addresses associated with the
Lightning node's underlying wallet. This allows for processing the
output of GetTransactions to determine which underlying wallet address a
corresponds to a transaction's outputs, as well as potentially many
other useful features and extensions.

#### Pull Request Checklist

- [x] If this is your first time contributing, we recommend you read the [Code
  Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md)
- [x] All changes are Go version 1.12 compliant
- [x] The code being submitted is commented according to [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation)
- [x] For new code: Code is accompanied by tests which exercise both
  the positive and negative (error paths) conditions (if applicable)
- [ ] For bug fixes: Code is accompanied by new tests which trigger
  the bug being fixed to prevent regressions
- [x] Any new logging statements use an appropriate subsystem and
  logging level
- [x] Code has been formatted with `go fmt`
- [x] Protobuf files (`lnrpc/**/*.proto`) have been formatted with
  `make rpc-format` and compiled with `make rpc`
- [ ] New configuration flags have been added to `sample-lnd.conf`
- [x] For code and documentation: lines are wrapped at 80 characters
  (the tab character should be counted as 8 characters, not 4, as some IDEs do
  per default)
- [x] Running `make check` does not fail any tests (Unable to run due to lack of local bitcoind setup) 
- [x] Running `go vet` does not report any issues
- [ ] Running `make lint` does not report any **new** issues that did not
  already exist
- [x] All commits build properly and pass tests. Only in exceptional
  cases it can be justifiable to violate this condition. In that case, the
  reason should be stated in the commit message.
- [x] Commits have a logical structure according to [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure)

### CODE REVIEW

**Review by jakesyl — COMMENTED** [review]
Needs a rebase

**Comment by jalavosus** [comment]
@jakesyl done.

File: go.sum [hunk 1]
```diff
@@ -348,20 +362,35 @@ golang.org/x/tools v0.0.0-20191029041327-9cc4af7d6b2c/go.mod h1:b+2E5dAYhXwXZwtn
...
+google.golang.org/grpc v1.27.0 h1:rRYRFMVgRv6E0D70Skyfsr28tDXIuuPZyWGMPdMcnXg=
+google.golang.org/grpc v1.27.0/go.mod h1:qbnxyOmOxrQa7FizSgH+ReBfzJrCY1pSN7KXBS8abTk=
+google.golang.org/protobuf v0.0.0-20200109180630-ec00e32a8dfd/go.mod h1:DFci5gLYBciE7Vtevhsrf46CRTquxDuWsQurQQe4oz8=
+google.golang.org/protobuf v0.0.0-20200221191635-4d8936d0db64/go.mod h1:kwYJMbMJ01Woi6D6+Kah6886xMZcty6N08ah7+eCXa0=
+google.golang.org/protobuf v0.0.0-20200228230310-ab0ca4ff8a60/go.mod h1:cfTl7dwQJ+fmap5saPgwCLgHXTUD7jkjRqWcaiX5VyM=
+google.golang.org/protobuf v1.20.1-0.20200309200217-e05f789c0967/go.mod h1:A+miEFZTKqfCUM6K7xSMQL9OKL/b6hQv+e19PK+JZNE=
+google.golang.org/protobuf v1.21.0/go.mod h1:47Nbq4nVaFHyn7ilMalzfO3qCViNmqZ2kzikPIcrTAo=
+google.golang.org/protobuf v1.22.0/go.mod h1:EGpADcykh3NcUnDUJcl1+ZksZNG86OlYog2l/sGQquU=
+google.golang.org/protobuf v1.23.1-0.20200526195155-81db48ad09cc/go.mod h1:EGpADcykh3NcUnDUJcl1+ZksZNG86OlYog2l/sGQquU=
+google.golang.org/protobuf v1.25.0/go.mod h1:9JNX74DMeImyA3h4bdi1ymwjUzf21/xIlbajtzgsN7c=
```
Comment (jakesyl) [inline go.sum:393]: `go mod tidy`?

**Review by jakesyl — APPROVED** [review]

File: go.sum (hunk 1 above)
Comment (jalavosus) [reply go.sum:393]: Sweet god that's messed up, yeah.

File: lntest/mock/walletcontroller.go [hunk 2]
```diff
@@ -66,6 +66,15 @@ func (w *WalletController) NewAddress(addrType lnwallet.AddressType,
 	return addr, nil
 }
 
+// ListWalletAddresses currently returns dummy values.
+func (w *WalletController) ListWalletAddresses(addrType lnwallet.AddressType) (
+	[]btcutil.Address, error) {
+
+	var walletAddresses []btcutil.Address
```
Comment (carlaKC) [inline lntest/mock/walletcontroller.go:73]: nit: don't need this var, could just return `[]btcutil.Address{}`

File: lnwallet/interface_test.go [hunk 3]
```diff
@@ -2533,6 +2533,80 @@ func testLastUnusedAddr(miner *rpctest.Harness,
...
+	}
+
+	for _, addrType := range addrTypes {
+		addr1, err := alice.LastUnusedAddress(addrType)
+		if err != nil {
+			t.Fatalf("unable to get addr: %v", err)
+		}
+
+		addrList1, err := alice.ListWalletAddresses(addrType)
+		if err != nil {
```
Comment (carlaKC) [inline lnwallet/interface_test.go:2556]: nit: we've started using the `require` package in unit tests, `require.NoError` will be useful in a few places here

File: lnwallet/interface_test.go [hunk 4]
```diff
@@ -2533,6 +2533,80 @@ func testLastUnusedAddr(miner *rpctest.Harness,
...
+			for _, walletAddr2 := range addrList2 {
+				if walletAddr1.String() == walletAddr2.String() {
+					hasExistingAddress = true
+					break CheckAddrListLoop
+				}
+			}
+		}
+
+		if !hasExistingAddress {
+			t.Fatalf("existing address before use and creation of new address not found in second call of wallet.ListWalletAddresses")
```
Comment (carlaKC) [inline lnwallet/interface_test.go:2605]: nit: wrap at 80

File: go.mod [hunk 5]
```diff
@@ -69,7 +69,8 @@ require (
 	golang.org/x/net v0.0.0-20191002035440-2ec189313ef0
 	golang.org/x/sys v0.0.0-20200202164722-d101bd2416d5
 	golang.org/x/time v0.0.0-20180412165947-fbb02b2291d2
-	google.golang.org/grpc v1.24.0
+	google.golang.org/grpc v1.27.0
```
Comment (carlaKC) [inline go.mod:72]: We intentionally stick to specific proto versions (see [readme](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/README.md#generate-protobuf-definitions)), so we don't need to bump any versions here.

File: lnrpc/autopilotrpc/autopilot.pb.go [hunk 6]
```diff
@@ -1,444 +1,726 @@
...
-	XXX_sizecache        int32    `json:"-"`
+	Active bool `protobuf:"varint,1,opt,name=active,proto3" json:"active,omitempty"`
 }
 
-func (m *StatusResponse) Reset()         { *m = StatusResponse{} }
-func (m *StatusResponse) String() string { return proto.CompactTextString(m) }
-func (*StatusResponse) ProtoMessage()    {}
-func (*StatusResponse) Descriptor() ([]byte, []int) {
-	return fileDescriptor_e0b9dc347a92e084, []int{1}
+func (x *StatusResponse) Reset() {
```
Comment (carlaKC) [inline lnrpc/autopilotrpc/autopilot.pb.go:79]: Most of these proto changes are unrelated to the change, reverting the version upgrade and regenerating will fix them.

File: lnrpc/walletrpc/walletkit.proto [hunk 7]
```diff
@@ -234,6 +239,13 @@ message AddrResponse {
     string addr = 1;
 }
 
+message AddrsResponse {
+    /*
+    Array of addresses encoded using a bech32 format.
+     */
```
Comment (carlaKC) [inline lnrpc/walletrpc/walletkit.proto:245]: nit: extra space here

File: lnrpc/walletrpc/walletkit_server.go [hunk 8]
```diff
@@ -445,6 +449,26 @@ func (w *WalletKit) NextAddr(ctx context.Context,
 	}, nil
 }
 
+// ListWalletAddresses returns a list of all addresses known to the underlying wallet.
+func (w *WalletKit) ListWalletAddresses(ctx context.Context,
+	req *AddrRequest) (*AddrsResponse, error) {
+
+	var addrsSlice []string
```
Comment (carlaKC) [inline lnrpc/walletrpc/walletkit_server.go:456]: nit: move this below `ListWalletAddresses` and preallocate `addrsSlice := make([]string, 0, len(walletAddrs))`

File: lnwallet/btcwallet/btcwallet.go [hunk 9]
```diff
@@ -290,6 +290,39 @@ func (b *BtcWallet) IsOurAddress(a btcutil.Address) bool {
...
+		keyScope = waddrmgr.KeyScopeBIP0084
+	case lnwallet.NestedWitnessPubKey:
+		keyScope = waddrmgr.KeyScopeBIP0049Plus
+	default:
+		return nil, fmt.Errorf("unknown address type")
+	}
+
+	walletAccounts, err := b.wallet.Accounts(keyScope)
+	if err != nil {
+		return walletAddresses, err
```
Comment (carlaKC) [inline lnwallet/btcwallet/btcwallet.go:311]: Could also just return `nil, err`, because the assumption is that the caller won't use the address set on error.

File: lnrpc/walletrpc/walletkit_server.go [hunk 10]
```diff
@@ -445,6 +449,26 @@ func (w *WalletKit) NextAddr(ctx context.Context,
...
+
+	walletAddrs, err := w.cfg.Wallet.ListWalletAddresses(lnwallet.WitnessPubKey)
```
Comment (carlaKC) [inline lnrpc/walletrpc/walletkit_server.go:458]: Why only `WitnessPubkey` here?

**Review by carlaKC — COMMENTED** [review]
Thanks for the PR, looking good so far. 

My main comment is that we don't want to bump the proto versions, so you'll need to reset the `go.mod` changes and regenerate.

File: lnrpc/walletrpc/walletkit.proto (hunk 7 above)
Comment (guggero) [reply lnrpc/walletrpc/walletkit.proto:245]: You can also run `make rpc-format` to automatically clean this up.

File: go.sum (hunk 1 above)
Comment (guggero) [reply go.sum:393]: Using the correct version of the RPC compilers (see https://github.com/lightningnetwork/lnd/blob/master/lnrpc/README.md#generate-protobuf-definitions as mentioned by Carla below) should also get rid of the diff in all unchanged `*.pb.go` files.

File: lnrpc/rest-annotations.yaml [hunk 11]
```diff
@@ -261,6 +261,9 @@ http:
     - selector: walletrpc.WalletKit.NextAddr
       post: "/v2/wallet/address/next"
       body: "*"
+    - selector: walletrpc.WalletKit.ListWalletAddresses
+      post: "/v2/wallet/addresses/list"
```
Comment (guggero) [inline lnrpc/rest-annotations.yaml:265]: REST URIs should be singular in my opinion, so this should just be `/v2/wallet/address/list`.
Also, because there are no input parameters, I think this can be a `get` method (the above `NextAddr` should be `get` as well, but we can't change that now as it would break existing users, but I'd like to at least have it correct for new RPCs).

File: lnrpc/walletrpc/walletkit.proto [hunk 12]
```diff
@@ -50,6 +50,11 @@ service WalletKit {
     */
     rpc NextAddr (AddrRequest) returns (AddrResponse);
 
+    /*
+    ListWalletAddresses returns all addresses in the underlying wallet.
+     */
+    rpc ListWalletAddresses (AddrRequest) returns (AddrsResponse);
```
Comment (guggero) [inline lnrpc/walletrpc/walletkit.proto:56]: Would be very useful to have this as a command in `lncli wallet` as well (see `cmd/lncli/walletrpc_active.go`).

**Comment by guggero** [comment]
Thanks for the PR, I think this is a very useful feature to have!
I'm going to give this a full review once you cleaned up the diff somewhat (see comments above).

One question I have is (haven't tested it myself yet): Which addresses does the underlying wallet return? All addresses up to the current index, including change addresses? Or just those the wallet thinks are being used because there is an associated transaction?

File: lnrpc/walletrpc/walletkit_server.go (hunk 10 above)
Comment (jalavosus) [reply lnrpc/walletrpc/walletkit_server.go:458]: @carlaKC I based that implementation detail off of the `NextAddress` command -- since that RPC only returns a WPK, it made sense to me to have this only return WPKs as well.

File: lnwallet/interface_test.go (hunk 3 above)
Comment (jalavosus) [reply lnwallet/interface_test.go:2556]: Can you describe an example behavior for this? I'll happily implement it, just need to see how

**Comment by jalavosus** [comment]
I'll have to test that behavior myself as well -- the behavior of btcwallet isn't _entirely_ clear to me, so I'll play around with it and see what it does give me

File: lnrpc/walletrpc/walletkit.proto (hunk 7 above)
Comment (jalavosus) [reply lnrpc/walletrpc/walletkit.proto:245]: Running `make rpc-format` doesn't seem to fix this -_- 

I'll just make the change manually

File: lnrpc/walletrpc/walletkit.proto (hunk 12 above)
Comment (jalavosus) [reply lnrpc/walletrpc/walletkit.proto:56]: Waiting for my local node to sync back up so I can test this, but I have it committed locally!

File: lnwallet/interface_test.go (hunk 3 above)
Comment (jalavosus) [reply lnwallet/interface_test.go:2556]: Ah, got it! It'll be in an incoming commit.

**Comment by jalavosus** [comment]
@guggero and @carlaKC, I just pushed (and rebased + force-pushed) a swath of changes related to your comments! Let me know what else I can do for this PR.

**Comment by jalavosus** [comment]
Just did a rebase onto master, and updated the `sendCoins()` test method in `lntest/harness.go` to use the non-deprecated `walletrpc.ListUnspent()` method in favor of the deprecated `lnrpc.ListUnspent()`.

**Comment by guggero** [comment]
You still have changes to some unrelated `*.pb.go` files in your diff. Please make sure you use the exact versions of `protoc` and the `protobuf` library installed as described here: https://github.com/lightningnetwork/lnd/blob/master/lnrpc/README.md#generate-protobuf-definitions
If you're on Linux, you can also just run `scripts/install_travis_proto.sh` to install everything correctly (will override any pre-installed `protoc`).

**Comment by jalavosus** [comment]
Time to break out my Linux VM; protoc on macOS and with go modules is proving to be a giant pain in the butt

**Comment by guggero** [comment]
Yes, it is a bit of a pain to pin those versions. There are ideas around dockerizing the proto compilation part, but that would require everyone to install Docker which is less light-weight than `protoc`.

**Comment by jakesyl** [comment]
@jalavosus I went ahead and applied the patch for this commit to xplorfin/lnd#10 so if you want to change the base branch to `xplorfin/listwallet` we should be good to go here

**Comment by Roasbeef** [comment]
The OG issue discussed just exposing `IsOurAddr` over the RPC interface, depending on the use case that may be more scalable than this call, since at times we use a very large look ahead, which would result in a large response for this proposed call, eventually requiring it to have proper pagination.

**Comment by carlaKC** [comment]
Closing due to inactivity.
